plugin/db/mysql/rollback.go

```diff
@@ -0,0 +1,139 @@
+package mysql
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// GetRollbackSQL generates the rollback SQL for the list of binlog events in the reversed order.
+func (txn BinlogTransaction) GetRollbackSQL(columnNames []string) (string, error) {
+	if len(txn) == 0 {
+		return "", nil
+	}
+	var sqlList []string
+	// Generate rollback SQL for each statement of the transaction in reversed order.
+	// Each statement may have multiple affected rows in a single binlog event. The order between them is irrelevant.
+	for i := len(txn) - 1; i >= 0; i-- {
+		e := txn[i]
+		if e.Type != WriteRowsEventType && e.Type != DeleteRowsEventType && e.Type != UpdateRowsEventType {
+			continue
+		}
+		sql, err := e.getRollbackSQL(columnNames)
+		if err != nil {
+			return "", err
+		}
+		sqlList = append(sqlList, sql)
+	}
+	return strings.Join(sqlList, "\n"), nil
+}
+
+func (e *BinlogEvent) getRollbackSQL(columnNames []string) (string, error) {
+	// 1. Remove the "### " prefix of each line.
+	// Prefix a "\n" to make the replacement easier.
+	body := strings.ReplaceAll("\n"+e.Body, "\n### ", "\n")
+
+	// 2. Switch "DELETE FROM" and "INSERT INTO".
+	// 3. Replace "WHERE" and "SET" with each other.
+	// 4. Replace "@i" with the column names.
+	// 5. Add a ";" at the end of each row.
+	var err error
+	switch e.Type {
+	case WriteRowsEventType:
+		body = strings.ReplaceAll(body, "\nINSERT INTO", "\nDELETE FROM")
+		// Trim the "\n" prefix we added at the first step.
+		body = strings.TrimPrefix(body, "\n")
+		body = strings.ReplaceAll(body, "\nSET", "\nWHERE")
+		body, err = replaceColumns(columnNames, body, "\nWHERE", " AND")
+		if err != nil {
+			return "", err
+		}
+		return addSemicolon(body, "\nDELETE FROM"), nil
+	case DeleteRowsEventType:
+		body = strings.ReplaceAll(body, "\nDELETE FROM", "\nINSERT INTO")
+		// Trim the "\n" prefix we added at the first step.
+		body = strings.TrimPrefix(body, "\n")
+		body = strings.ReplaceAll(body, "\nWHERE", "\nSET")
+		body, err = replaceColumns(columnNames, body, "\nSET", ",")
+		if err != nil {
+			return "", err
+		}
+		return addSemicolon(body, "\nINSERT INTO"), nil
+	case UpdateRowsEventType:
+		// Trim the "\n" prefix we added at the first step.
+		body = strings.TrimPrefix(body, "\n")
+		body = strings.ReplaceAll(body, "\nWHERE", "\nOLDWHERE")
```

Comment:
We could do something better than this, but. Let's refactor it to a better way later.